File: examples/python/run_async_model_wait.py
import numpy as np
import sys
from dx_engine import InferenceEngine

import queue
import threading

q = queue.Queue()


def inferenceThreadFunc(ie, loopCount):

    count = 0

    while(True):
    
        # pop item from queue 
        jobId = q.get()

        # waiting for the inference to complete by jobId
        # ownership of the outputs is transferred to the user 
        outputs = ie.Wait(jobId)

        # post processing
        # postProcessing(outputs);

        # something to do


        print("Inference outputs corresponding to jobId=", jobId, "index=", count)

        count += 1
        if ( count >= loopCount ):
            break
   
    return 0


if __name__ == "__main__":
    DEFAULT_LOOP_COUNT = 1
    loop_count = DEFAULT_LOOP_COUNT
    modelPath = ""
    argc = len(sys.argv)
    if ( argc > 1 ) :
        modelPath = sys.argv[1];
        if ( argc > 2 ) :
            loop_count = int(sys.argv[2])
    else:
        print("[Usage] run_async_model [dxnn-file-path] [loop-count]")
        exit(-1)
    

    # create inference engine instance with model
    ie = InferenceEngine(modelPath)

    # do not register call back function
    # ie.register_callback(onInferenceCallbackFunc)

    t1 = threading.Thread(target=inferenceThreadFunc, args=(ie, loop_count))

    t1.start()

    input = [np.zeros(ie.input_size(), dtype=np.uint8)]

    # inference loop
    for i in range(loop_count):

        
        # inference asynchronously, use all npu cores
        # if device-load >= max-load-value, this function will block  
        jobId = ie.RunAsync(input, user_arg=0)

        q.put(jobId)

        print("Inference start (async)", i)

    t1.join()

    exit(0)

File: examples/run_async_model_wait/run_async_model_wait.cpp
// Copyright (c) 2022 DEEPX Corporation. All rights reserved.
// Licensed under the MIT License.

#include "dxrt/dxrt_api.h"
#include "concurrent_queue.h"

#include <string>
#include <iostream>


// concurrent queue is a thread-safe queue data structure 
// designed to be used in a multi-threaded environment
static ConcurrentQueue<int> gJobIdQueue(32);

// user thread to wait for the completion of inference 
static int inferenceThreadFunc(dxrt::InferenceEngine& ie, int loopCount)
{
    int count = 0;

    while(true)
    {
        // pop item from queue 
        int jobId = gJobIdQueue.pop();

        try
        {
            // waiting for the inference to complete by jobId
            // ownership of the outputs is transferred to the user 
            auto outputs = ie.Wait(jobId);

            // post processing
            // postProcessing(outputs);
            (void)outputs;

            // something to do
        }
        catch (const dxrt::Exception& e)
        {
            std::cerr << e.what() << " error-code=" << e.code() << std::endl;
            return -1;
        }

        std::cout << "Inference outputs corresponding to jobId(" << jobId << ")" << std::endl;

        count++;
        if ( count >= loopCount ) break;

    } // while

    return 0;
}

int main(int argc, char* argv[])
{
    const int DEFAULT_LOOP_COUNT = 1;
    
    std::string modelPath;
    int loop_count = DEFAULT_LOOP_COUNT;
    if ( argc > 1 )
    {
        modelPath = argv[1];

        if ( argc > 2 ) 
        {
            loop_count = std::stoi(argv[2]);
        }

    }
    else
    {
        std::cout << "[Usage] run_async_model_wait [dxnn-file-path] [loop-count]" << std::endl;
        return -1;
    }


    try
    {
        // create inference engine instance with model
        dxrt::InferenceEngine ie(modelPath);

        // do not register call back function
        // inferenceEngine.RegisterCallBack(onInferenceCallbackFunc);

        // create thread
        auto t1 = std::thread(inferenceThreadFunc, std::ref(ie), loop_count);

        // create temporary input buffer for example
        std::vector<uint8_t> inputPtr(ie.input_size(), 0);

        auto start = std::chrono::high_resolution_clock::now();

        // inference loop
        for(int i = 0; i < loop_count; ++i)
        {
            // no need user argument
            // UserData *userData = getUserDataInstanceFromDataPool();

            // inference asynchronously, use all npu cores
            // if device-load >= max-load-value, this function will block
            auto jobId = ie.RunAsync(inputPtr.data());

            // push jobId in global queue variable
            gJobIdQueue.push(jobId);

            std::cout << "Inference request submitted with jobId(" << jobId << ")" << std::endl;

        } // for i

        t1.join();

        auto end = std::chrono::high_resolution_clock::now();
        std::chrono::duration<double, std::milli> duration = end - start;

        double total_time = duration.count();
        double avg_latency = total_time / static_cast<double>(loop_count);
        double fps = 1000.0 / avg_latency;

        std::cout << "-----------------------------------" << std::endl;
        std::cout << "Total Time: " << total_time << " ms" << std::endl;
        std::cout << "Average Latency: " << avg_latency << " ms" << std::endl;
        std::cout << "FPS: " << fps << " frames/sec" << std::endl;
        std::cout << "Success" << std::endl;
        std::cout << "-----------------------------------" << std::endl;
    }
    catch (const dxrt::Exception& e)
    {
        std::cerr << e.what() << " error-code=" << e.code() << std::endl;
        return -1;
    }
    catch (const std::exception& e)
    {
        std::cerr << e.what() << std::endl;
        return -1;
    }
    catch(...)
    {
        std::cerr << "Exception" << std::endl;
        return -1;
    }
    
    return 0;
}

